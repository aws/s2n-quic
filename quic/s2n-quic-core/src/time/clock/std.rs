// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use ::std::time::Instant;

impl<C: 'static + Clock> Clock for &'static ::std::thread::LocalKey<C> {
    fn get_time(&self) -> Timestamp {
        self.with(|clock| clock.get_time())
    }
}

#[derive(Clone, Copy, Debug)]
pub struct StdClock {
    epoch: Instant,
}

impl Default for StdClock {
    fn default() -> Self {
        Self {
            epoch: Instant::now(),
        }
    }
}

impl StdClock {
    /// Creates a new `StdClock` with the given epoch
    pub const fn new(epoch: Instant) -> Self {
        Self { epoch }
    }
}

impl Clock for StdClock {
    fn get_time(&self) -> Timestamp {
        unsafe { Timestamp::from_duration(self.epoch.elapsed()) }
    }
}

#[test]
#[cfg_attr(miri, ignore)] // time isn't queryable in miri
fn monotonicity_test() {
    let clock = StdClock::default();
    let ts1 = clock.get_time();
    ::std::thread::sleep(Duration::from_millis(50));
    let ts2 = clock.get_time();
    assert!(ts2 - ts1 >= Duration::from_millis(50));
}
