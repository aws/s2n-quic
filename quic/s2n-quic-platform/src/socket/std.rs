// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    buffer::Buffer,
    message::{
        queue,
        simple::{self, Message, Ring},
        Message as _,
    },
};
use errno::errno;
use s2n_quic_core::{event, inet::SocketAddress, path::LocalAddress};

pub use simple::Handle;

pub trait Socket {
    type Error: Error;

    /// Receives a payload and returns the length and source address
    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, Option<SocketAddress>), Self::Error>;

    /// Sends a payload to the given address and returns the length of the sent payload
    fn send_to(&self, buf: &[u8], addr: &SocketAddress) -> Result<usize, Self::Error>;
}

#[cfg(feature = "std")]
impl Socket for std::net::UdpSocket {
    type Error = std::io::Error;

    fn recv_from(&self, buf: &mut [u8]) -> Result<(usize, Option<SocketAddress>), Self::Error> {
        debug_assert!(!buf.is_empty());
        let (len, addr) = self.recv_from(buf)?;
        Ok((len, Some(addr.into())))
    }

    fn send_to(&self, buf: &[u8], addr: &SocketAddress) -> Result<usize, Self::Error> {
        debug_assert!(!buf.is_empty());
        let addr: std::net::SocketAddr = (*addr).into();
        self.send_to(buf, &addr)
    }
}

pub trait Error {
    fn would_block(&self) -> bool;
    fn was_interrupted(&self) -> bool;
    fn permission_denied(&self) -> bool;
}

#[cfg(feature = "std")]
impl Error for std::io::Error {
    fn would_block(&self) -> bool {
        self.kind() == std::io::ErrorKind::WouldBlock
    }

    fn was_interrupted(&self) -> bool {
        self.kind() == std::io::ErrorKind::Interrupted
    }

    fn permission_denied(&self) -> bool {
        self.kind() == std::io::ErrorKind::PermissionDenied
    }
}

#[derive(Debug, Default)]
pub struct Queue<B: Buffer>(queue::Queue<Ring<B>>);

impl<B: Buffer> Queue<B> {
    pub fn new(buffer: B) -> Self {
        let queue = queue::Queue::new(Ring::new(buffer, 1));

        Self(queue)
    }

    pub fn free_len(&self) -> usize {
        self.0.free_len()
    }

    pub fn occupied_len(&self) -> usize {
        self.0.occupied_len()
    }

    pub fn set_local_address(&mut self, local_address: LocalAddress) {
        self.0.set_local_address(local_address)
    }

    pub fn tx<S: Socket, Publisher: event::EndpointPublisher>(
        &mut self,
        socket: &S,
        publisher: &mut Publisher,
    ) -> Result<usize, S::Error> {
        let mut count = 0;
        let mut entries = self.0.occupied_mut();

        for entry in entries.as_mut() {
            if let Some(remote_address) = entry.remote_address() {
                match socket.send_to(entry.payload_mut(), &remote_address) {
                    Ok(_) => {
                        count += 1;

                        publisher.on_platform_tx(event::builder::PlatformTx { count: 1 });
                    }
                    Err(err) if count > 0 && err.would_block() => {
                        break;
                    }
                    Err(err) if err.was_interrupted() || err.permission_denied() => {
                        break;
                    }
                    Err(err) => {
                        entries.finish(count);

                        publisher.on_platform_rx_error(event::builder::PlatformRxError {
                            errno: errno().0,
                        });

                        return Err(err);
                    }
                }
            }
        }

        entries.finish(count);

        Ok(count)
    }

    pub fn rx<S: Socket, Publisher: event::EndpointPublisher>(
        &mut self,
        socket: &S,
        publisher: &mut Publisher,
    ) -> Result<usize, S::Error> {
        let mut count = 0;
        let mut entries = self.0.free_mut();

        while let Some(entry) = entries.get_mut(count) {
            match socket.recv_from(entry.payload_mut()) {
                Ok((payload_len, Some(remote_address))) => {
                    entry.set_remote_address(&remote_address);
                    unsafe {
                        // Safety: The payload_len should not be bigger than the number of
                        // allocated bytes.

                        debug_assert!(payload_len < entry.payload_len());
                        let payload_len = payload_len.min(entry.payload_len());

                        entry.set_payload_len(payload_len);
                    }

                    count += 1;

                    publisher.on_platform_rx(event::builder::PlatformRx { count: 1 });
                }
                Ok((_payload_len, None)) => {}
                Err(err) if count > 0 && err.would_block() => {
                    break;
                }
                Err(err) if err.was_interrupted() => {
                    break;
                }
                Err(err) => {
                    entries.finish(count);

                    publisher
                        .on_platform_rx_error(event::builder::PlatformRxError { errno: errno().0 });

                    return Err(err);
                }
            }
        }

        entries.finish(count);

        Ok(count)
    }

    pub fn rx_queue(&mut self) -> queue::OccupiedWipe<Message> {
        self.0.occupied_wipe_mut()
    }

    pub fn tx_queue(&mut self) -> queue::Free<Message> {
        self.0.free_mut()
    }
}
