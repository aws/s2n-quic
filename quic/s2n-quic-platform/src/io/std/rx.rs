// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    buffer,
    message::{
        queue,
        simple::{Message, Ring},
        Message as MessageTrait,
    },
    socket,
};
use s2n_quic_core::io::rx;
use std::io;

impl_io!(Rx);
impl_socket_raw_delegate!(
    impl[Buffer: buffer::Buffer, Socket: socket::raw::AsRaw] Rx<Buffer, Socket>,
    |self| &self.socket
);
impl_socket_mio_delegate!(
    impl[Buffer: buffer::Buffer, Socket: mio::Evented] Rx<Buffer, Socket>,
    |self| &self.socket
);

impl<
        'a,
        Buffer: buffer::Buffer,
        Socket: socket::Simple<Error = io::Error> + socket::Socket<Error = io::Error>,
    > rx::Rx<'a> for Rx<Buffer, Socket>
{
    type Queue = queue::OccupiedWipe<'a, Message>;
    type Error = io::Error;

    fn queue(&'a mut self) -> Self::Queue {
        self.queue.occupied_wipe_mut()
    }

    fn len(&self) -> usize {
        self.queue.free_len()
    }

    fn poll_receive(
        &mut self,
        cx: &mut core::task::Context<'_>,
    ) -> core::task::Poll<io::Result<usize>> {
        let Self { socket, queue } = self;
        socket.poll_receive(cx, |socket| {
            let mut count = 0;
            let mut free = queue.free_mut();

            while let Some(entry) = free.get_mut(count) {
                match socket.recv_from(entry.payload_mut()) {
                    Ok((payload_len, Some(remote_address))) => {
                        entry.set_remote_address(&remote_address);
                        unsafe {
                            // Safety: The payload_len should not be bigger than the number of
                            // allocated bytes.

                            debug_assert!(payload_len < entry.payload_len());
                            let payload_len = payload_len.min(entry.payload_len());

                            entry.set_payload_len(payload_len);
                        }
                        count += 1;
                    }
                    Ok((_payload_len, None)) => {}
                    Err(err) => {
                        if count > 0 && err.kind() == io::ErrorKind::WouldBlock {
                            break;
                        } else {
                            free.finish(count);
                            return Err(err);
                        }
                    }
                }
            }

            free.finish(count);

            Ok(count)
        })
    }
}

#[cfg(test)]
mod tests {
    impl_io_rx_tests!();
}
