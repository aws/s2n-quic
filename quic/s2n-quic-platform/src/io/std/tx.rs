use crate::{
    buffer,
    message::{
        queue,
        simple::{Message, Ring},
        Message as MessageTrait,
    },
    socket,
};
use s2n_quic_core::io::tx;
use std::io;

impl_io!(Tx);
impl_socket_raw_delegate!(
    impl[Buffer: buffer::Buffer, Socket: socket::raw::AsRaw] Tx<Buffer, Socket>,
    |self| &self.socket
);
impl_socket_mio_delegate!(
    impl[Buffer: buffer::Buffer, Socket: mio::Evented] Tx<Buffer, Socket>,
    |self| &self.socket
);

impl<
        'a,
        Buffer: buffer::Buffer,
        Socket: socket::Simple<Error = io::Error> + socket::Socket<Error = io::Error>,
    > tx::Tx<'a> for Tx<Buffer, Socket>
{
    type Queue = queue::Free<'a, Message>;
    type Error = io::Error;

    fn queue(&'a mut self) -> Self::Queue {
        self.queue.free_mut()
    }

    fn len(&self) -> usize {
        self.queue.occupied_len()
    }

    fn poll_transmit(
        &mut self,
        cx: &mut core::task::Context<'_>,
    ) -> core::task::Poll<io::Result<usize>> {
        let Self { socket, queue } = self;
        socket.poll_transmit(cx, |socket| {
            let mut count = 0;
            let mut occupied = queue.occupied_mut();

            for entry in occupied.as_mut() {
                if let Some(remote_address) = entry.remote_address() {
                    match socket.send_to(entry.payload_mut(), &remote_address) {
                        Ok(_) => {
                            count += 1;
                        }
                        Err(err) => {
                            if count > 0 && err.kind() == io::ErrorKind::WouldBlock {
                                break;
                            } else {
                                occupied.finish(count);
                                return Err(err);
                            }
                        }
                    }
                }
            }

            occupied.finish(count);

            Ok(count)
        })
    }
}

#[cfg(test)]
mod tests {
    impl_io_tx_tests!();
}
