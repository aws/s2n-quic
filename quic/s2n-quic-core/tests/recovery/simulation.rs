use core::{fmt, ops::Range, time::Duration};
use insta::assert_debug_snapshot;
use plotters::prelude::*;
use s2n_quic_core::{
    packet::number::PacketNumberSpace,
    path::MINIMUM_MTU,
    recovery::{CongestionController, CubicCongestionController, RTTEstimator},
    time::{Clock, NoopClock, Timestamp},
};
use std::{
    env,
    path::{Path, PathBuf},
};

const CHART_DIMENSIONS: (u32, u32) = (1024, 768);

// These simulations are too slow for Miri
#[test]
#[cfg_attr(miri, ignore)]
fn slow_start_unlimited_test() {
    let cc = CubicCongestionController::new(MINIMUM_MTU);

    slow_start_unlimited(cc, 12).finish();
}

#[test]
#[cfg_attr(miri, ignore)]
fn loss_at_3mb_test() {
    let cc = CubicCongestionController::new(MINIMUM_MTU);

    loss_at_3mb(cc, 135).finish();
}

#[derive(Debug)]
struct Simulation {
    name: &'static str,
    cc: &'static str,
    rounds: Vec<Round>,
}

struct Round {
    number: usize,
    cwnd: u32,
}

impl fmt::Debug for Round {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:>3}: cwnd: {}", self.number, self.cwnd)
    }
}

impl Simulation {
    fn finish(&self) {
        if let Ok(dir) = env::var("RECOVERY_SIM_DIR") {
            let mut path = PathBuf::new();
            path.push(dir);
            path.push(self.filename());
            path.set_extension("svg");
            self.plot(&path);
        } else {
            self.assert_snapshot();
        }
    }

    fn plot<T: AsRef<Path> + ?Sized>(&self, path: &T) {
        let root_area = SVGBackend::new(path, CHART_DIMENSIONS).into_drawing_area();
        root_area.fill(&WHITE).expect("Could not fill chart");

        let mut ctx = ChartBuilder::on(&root_area)
            .set_label_area_size(LabelAreaPosition::Left, 100)
            .set_label_area_size(LabelAreaPosition::Bottom, 60)
            .margin(20)
            .caption(self.name(), ("sans-serif", 40))
            .build_cartesian_2d(self.x_spec(), self.y_spec())
            .expect("Could not build chart");

        ctx.configure_mesh()
            .x_desc("Transmission Round")
            .label_style(("sans-serif", 20))
            .y_desc("Congestion window size (bytes)")
            .draw()
            .expect("Could not configure mesh");

        ctx.draw_series(LineSeries::new(
            self.rounds.iter().map(|x| (x.number as i32, x.cwnd as i32)),
            &GREEN,
        ))
        .expect("Could not draw series");
    }

    fn x_spec(&self) -> Range<i32> {
        0..(self.rounds.len() as i32 + 1)
    }

    fn y_spec(&self) -> Range<i32> {
        let max = self.rounds.iter().map(|r| r.cwnd as i32).max().unwrap_or(0);

        0..max + MINIMUM_MTU as i32
    }

    fn assert_snapshot(&self) {
        assert_debug_snapshot!(self.filename(), self);
    }

    fn name(&self) -> String {
        let mut name = String::new();
        name.push_str(self.name);
        name.push_str(" - ");
        name.push_str(self.cc.split("::").last().unwrap());
        name
    }

    fn filename(&self) -> String {
        self.name().split_whitespace().collect()
    }
}

/// Simulates a network with no congestion experienced
fn slow_start_unlimited<CC: CongestionController>(
    mut congestion_controller: CC,
    num_rounds: usize,
) -> Simulation {
    let mut bytes_in_flight = 0;
    let time_zero = NoopClock.get_time();
    let mut rtt_estimator = RTTEstimator::new(Duration::from_millis(0));
    let mut rounds = Vec::with_capacity(num_rounds);
    let packet_size = MINIMUM_MTU as usize;

    for round in 0..num_rounds {
        rounds.push(Round {
            number: round,
            cwnd: congestion_controller.congestion_window(),
        });

        let sent_time = time_zero + Duration::from_millis(100 * round as u64);

        // Send the full congestion window of bytes
        while bytes_in_flight <= congestion_controller.congestion_window() {
            congestion_controller.on_packet_sent(sent_time, packet_size);
            bytes_in_flight += packet_size as u32;
        }

        let ack_receive_time = sent_time + Duration::from_millis(50);

        // Ack the full congestion window of bytes
        while bytes_in_flight > 0 {
            congestion_controller.on_packet_ack(
                ack_receive_time,
                packet_size,
                &rtt_estimator,
                ack_receive_time,
            );
            bytes_in_flight -= packet_size as u32;
        }

        rtt_estimator.update_rtt(
            Duration::from_millis(0),
            ack_receive_time - sent_time,
            ack_receive_time,
            true,
            PacketNumberSpace::ApplicationData,
        );
    }

    Simulation {
        name: "Slow Start Unlimited",
        cc: core::any::type_name::<CC>(),
        rounds,
    }
}

/// Simulates a network that experienced loss at a 3MB congestion window
fn loss_at_3mb<CC: CongestionController>(
    mut congestion_controller: CC,
    num_rounds: usize,
) -> Simulation {
    let time_zero = NoopClock.get_time();
    let mut rtt_estimator = RTTEstimator::new(Duration::from_millis(0));
    let mut rounds = Vec::with_capacity(num_rounds);

    // Ensure the congestion window is fully utilized
    congestion_controller.on_packet_sent(time_zero, u32::MAX as usize);

    let mut ack_receive_time = time_zero + Duration::from_millis(1);

    // Update the rtt with 200 ms
    rtt_estimator.update_rtt(
        Duration::from_millis(0),
        Duration::from_millis(200),
        time_zero,
        true,
        PacketNumberSpace::ApplicationData,
    );

    let mut slow_start_round = 0;

    while congestion_controller.congestion_window() < 3_000_000 && slow_start_round < num_rounds {
        ack_receive_time += Duration::from_millis(200);

        // Ack the full congestion window
        ack_cwnd(&mut congestion_controller, &rtt_estimator, ack_receive_time);

        rounds.push(Round {
            number: slow_start_round,
            cwnd: congestion_controller.congestion_window(),
        });

        slow_start_round += 1;
    }

    // Lose a packet to exit slow start
    congestion_controller.on_packets_lost(MINIMUM_MTU as u32, false, ack_receive_time);

    for round in slow_start_round..num_rounds {
        rounds.push(Round {
            number: round,
            cwnd: congestion_controller.congestion_window(),
        });

        ack_receive_time += Duration::from_millis(200);

        // Ack the full congestion window
        ack_cwnd(&mut congestion_controller, &rtt_estimator, ack_receive_time);
    }

    Simulation {
        name: "Loss at 3MB",
        cc: core::any::type_name::<CC>(),
        rounds,
    }
}

/// Acknowledge a full congestion window of packets using the given congestion controller
fn ack_cwnd<CC: CongestionController>(
    congestion_controller: &mut CC,
    rtt_estimator: &RTTEstimator,
    timestamp: Timestamp,
) {
    let mut cwnd = congestion_controller.congestion_window();
    while cwnd >= MINIMUM_MTU as u32 {
        congestion_controller.on_packet_ack(
            timestamp,
            MINIMUM_MTU as usize,
            rtt_estimator,
            timestamp,
        );
        cwnd -= MINIMUM_MTU as u32;
        // Ensure the congestion window is always fully utilized by sending a packet the
        // same size as the one that we just acked.
        congestion_controller.on_packet_sent(timestamp, MINIMUM_MTU as usize);
    }
}
