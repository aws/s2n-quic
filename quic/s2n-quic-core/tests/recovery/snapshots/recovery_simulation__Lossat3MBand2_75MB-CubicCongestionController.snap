---
source: quic/s2n-quic-core/tests/recovery/simulation.rs
expression: self

---
Simulation {
    name: "Loss at 3MB and 2.75MB",
    description: "Loss encountered at ~3MB and ~2.75MB",
    cc: "s2n_quic_core::recovery::cubic::CubicCongestionController",
    rounds: [
          0: cwnd: 12000,
          1: cwnd: 24000,
          2: cwnd: 48000,
          3: cwnd: 96000,
          4: cwnd: 192000,
          5: cwnd: 384000,
          6: cwnd: 768000,
          7: cwnd: 1536000,
          8: cwnd: 3072000,
          9: cwnd: 2150400,
         10: cwnd: 2177957,
         11: cwnd: 2214787,
         12: cwnd: 2254221,
         13: cwnd: 2293776,
         14: cwnd: 2332539,
         15: cwnd: 2370177,
         16: cwnd: 2406593,
         17: cwnd: 2441748,
         18: cwnd: 2475647,
         19: cwnd: 2508321,
         20: cwnd: 2539773,
         21: cwnd: 2570031,
         22: cwnd: 2599120,
         23: cwnd: 2627060,
         24: cwnd: 2653885,
         25: cwnd: 2679601,
         26: cwnd: 2704239,
         27: cwnd: 2727827,
         28: cwnd: 2750384,
         29: cwnd: 1925268,
         30: cwnd: 1941343,
         31: cwnd: 1962657,
         32: cwnd: 1985264,
         33: cwnd: 2007707,
         34: cwnd: 2029461,
         35: cwnd: 2050340,
         36: cwnd: 2070290,
         37: cwnd: 2089312,
         38: cwnd: 2107414,
         39: cwnd: 2124616,
         40: cwnd: 2140939,
         41: cwnd: 2156407,
         42: cwnd: 2171042,
         43: cwnd: 2184867,
         44: cwnd: 2197904,
         45: cwnd: 2210180,
         46: cwnd: 2221715,
         47: cwnd: 2232534,
         48: cwnd: 2242655,
         49: cwnd: 2252104,
         50: cwnd: 2260906,
         51: cwnd: 2269084,
         52: cwnd: 2276661,
         53: cwnd: 2283661,
         54: cwnd: 2290099,
         55: cwnd: 2296005,
         56: cwnd: 2301409,
         57: cwnd: 2306313,
         58: cwnd: 2310764,
         59: cwnd: 2314767,
         60: cwnd: 2318357,
         61: cwnd: 2321546,
         62: cwnd: 2324376,
         63: cwnd: 2326842,
         64: cwnd: 2329004,
         65: cwnd: 2330839,
         66: cwnd: 2332399,
         67: cwnd: 2333730,
         68: cwnd: 2334786,
         69: cwnd: 2335635,
         70: cwnd: 2336315,
         71: cwnd: 2336866,
         72: cwnd: 2337224,
         73: cwnd: 2337412,
         74: cwnd: 2337520,
         75: cwnd: 2337569,
         76: cwnd: 2337582,
         77: cwnd: 2337583,
         78: cwnd: 2337595,
         79: cwnd: 2337640,
         80: cwnd: 2337742,
         81: cwnd: 2337923,
         82: cwnd: 2338208,
         83: cwnd: 2338618,
         84: cwnd: 2339141,
         85: cwnd: 2339769,
         86: cwnd: 2340560,
         87: cwnd: 2341587,
         88: cwnd: 2342878,
         89: cwnd: 2344370,
         90: cwnd: 2346137,
         91: cwnd: 2348230,
         92: cwnd: 2350614,
         93: cwnd: 2353349,
         94: cwnd: 2356457,
         95: cwnd: 2359942,
         96: cwnd: 2363850,
         97: cwnd: 2368184,
         98: cwnd: 2372983,
         99: cwnd: 2378258,
        100: cwnd: 2384036,
        101: cwnd: 2390348,
        102: cwnd: 2397201,
        103: cwnd: 2404634,
        104: cwnd: 2412660,
        105: cwnd: 2421304,
        106: cwnd: 2430592,
        107: cwnd: 2440548,
        108: cwnd: 2451186,
        109: cwnd: 2462537,
        110: cwnd: 2474626,
        111: cwnd: 2487472,
        112: cwnd: 2501092,
        113: cwnd: 2515519,
        114: cwnd: 2530772,
        115: cwnd: 2546871,
        116: cwnd: 2563849,
        117: cwnd: 2581720,
        118: cwnd: 2600508,
        119: cwnd: 2620239,
    ],
}
